test/langtools/jdk/javadoc/doclet/testIndex/TestSelfIndexing.java

```diff
@@ -0,0 +1,165 @@
+                                    .map(link -> new PathAndString(pac.path, link)))
+                            .findAny();
+                    r.ifPresentOrElse(p -> failed(p.toString()), () -> passed(t.formatted(l)));
+                }
+                i++;
+            }
+        }
+    }
+
+    // ----------- support and infrastructure -----------
+
+    private static Stream<Path> findIndexFiles(Path start) throws IOException {
+        return Files.find(start, Integer.MAX_VALUE, (path, attr) -> {
+            if (attr.isDirectory())
+                return false;
+            var fileName = path.getFileName().toString();
+            if (!fileName.endsWith(".html") && !fileName.endsWith(".js"))
+                return false;
+            if (!fileName.contains("-index") && !fileName.contains("index-"))
+                return false;
+            var underDocFiles = StreamSupport.stream(Spliterators.spliterator(path.iterator(),
+                            Integer.MAX_VALUE, Spliterator.ORDERED), false)
+                    .anyMatch(p -> p.equals(DOC_FILES));
+            return !underDocFiles;
+        });
+    }
+
+    private static final Path DOC_FILES = Path.of("doc-files");
+
+    // good enough to capture relevant parts of URLs that javadoc uses,
+    // from html and js files alike
+    private static final Pattern URL = Pattern.compile(
+            "(?<path>([a-zA-Z.%0-9-]+/)*+)(?<file>[a-zA-Z.%0-9-]+\\.html)#[a-zA-Z.%0-9-]+");
+
+    static {
+        assert findLinksToDerivedPages("module-summary.html#a").findAny().isEmpty();
+        assert findLinksToDerivedPages("package-summary.html#a").findAny().isEmpty();
+        assert findLinksToDerivedPages("Exception.html#a").findAny().isEmpty();
+        assert findLinksToDerivedPages("util/doc-files/coll-index.html#a").findAny().isEmpty();
+        assert findLinksToDerivedPages("util/doc-files/index-all.html#a").findAny().isEmpty(); // tricky
+
+
+        assert findLinksToDerivedPages("index-all.html#a").findAny().isPresent();
+        assert findLinksToDerivedPages("index-17.html#a").findAny().isPresent();
+    }
+
+    // NOTE: this will not find self-links that are allowed on some index pages.
+    // For example, the quick-jump first-character links, such as #I:A,
+    // #I:B, etc., on the top and at the bottom of index-all.html
+    private static Stream<String> findLinksToDerivedPages(String content) {
+        return URL.matcher(content).results()
+                .filter(r -> {
+                    String f = r.group("file");
+                    switch (f) {
+                        case "index-all.html", "deprecated-list.html", "overview-tree.html",
+                                "package-use.html", "package-tree.html", "preview-list.html",
+                                "new-list.html", "allclasses-index.html", "allpackages-index.html",
+                                "constant-values.html", "system-properties.html", "serialized-form.html"
```

Comment:
The hard-coded list of derived HTML files could become outdated when new lists are added. For example, we just recently added `restricted-list.html` for restricted methods. Could we detect this from the presence of a dash in the file name, as this is our way to avoid conflict of derived files with Java types?

Reply:
While not every file with a hyphen in its name is derived, we might be able to future-proof and shorten this check by somewhat inverting it:

    .filter(r -> {
        String f = r.group("file");
        if (!f.contains("-"))
            return false;
        return switch (f) {
            case "package-summary.html",
                    "module-summary.html",
                    "overview-summary.html",
                    "help-doc.html" -> false;
            default -> {
                String p = r.group("path");
                yield !p.contains("/doc-files/") && !p.startsWith("doc-files/");
            }
        };
    })

@hns?